packages/storage-plus/src/prefix.rs

```diff
@@ -8,40 +8,49 @@ use std::ops::Deref;
 
 use crate::de::KeyDeserialize;
 use crate::helpers::{namespaces_with_key, nested_namespaces_with_key};
-use crate::int_key::CwIntKey;
 use crate::iter_helpers::{concat, deserialize_kv, deserialize_v, trim};
 use crate::keys::Key;
-use crate::{Endian, Prefixer};
+use crate::{Prefixer, PrimaryKey};
 
-/// Bound is used to defines the two ends of a range, more explicit than Option<u8>
-/// None means that we don't limit that side of the range at all.
-/// Include means we use the given bytes as a limit and *include* anything at that exact key
-/// Exclude means we use the given bytes as a limit and *exclude* anything at that exact key
+/// `RawBound` is used to define the two ends of a range, more explicit than `Option<u8>`.
+/// `None` means that we don't limit that side of the range at all.
+/// `Inclusive` means we use the given bytes as a limit and *include* anything at that exact key.
+/// `Exclusive` means we use the given bytes as a limit and *exclude* anything at that exact key.
+/// See `Bound` for a type safe way to build these bounds.
 #[derive(Clone, Debug)]
-pub enum Bound {
+pub enum RawBound {
     Inclusive(Vec<u8>),
     Exclusive(Vec<u8>),
 }
 
-impl Bound {
-    /// Turns optional binary, like Option<CanonicalAddr> into an inclusive bound
-    pub fn inclusive<T: Into<Vec<u8>>>(limit: T) -> Self {
-        Bound::Inclusive(limit.into())
-    }
+/// `Bound` is used to define the two ends of a range.
+/// `None` means that we don't limit that side of the range at all.
+/// `Inclusive` means we use the given value as a limit and *include* anything at that exact key.
+/// `Exclusive` means we use the given value as a limit and *exclude* anything at that exact key.
+#[derive(Clone, Debug)]
+pub enum Bound<'a, K: PrimaryKey<'a>> {
+    Inclusive((K, PhantomData<&'a bool>)),
+    Exclusive((K, PhantomData<&'a bool>)),
+    InclusiveRaw(Vec<u8>),
+    ExclusiveRaw(Vec<u8>),
+}
 
-    /// Turns optional binary, like Option<CanonicalAddr> into an exclusive bound
-    pub fn exclusive<T: Into<Vec<u8>>>(limit: T) -> Self {
-        Bound::Exclusive(limit.into())
+impl<'a, K: PrimaryKey<'a>> Bound<'a, K> {
+    pub fn inclusive<T: Into<K>>(k: T) -> Self {
+        Self::Inclusive((k.into(), PhantomData))
     }
 
-    /// Turns an int, like Option<u32> into an inclusive bound
-    pub fn inclusive_int<T: CwIntKey + Endian>(limit: T) -> Self {
-        Bound::Inclusive(limit.to_cw_bytes().into())
+    pub fn exclusive<T: Into<K>>(k: T) -> Self {
+        Self::Exclusive((k.into(), PhantomData))
     }
```

Comment:
I wonder, wouldn't it make sense to include additional `inclusive_raw` and `exclusive_raw` helpers, which would contain those `.into()` inside implementations? I think it could be used in couple places at top of this PR.

Reply:
I thought about that, and decided not to. On the one hand, building these is really simple. On the other, it's better to discourage using these. If you need them, it's probably because you are doing something dodgy.